### PR TITLE
Combined infill max layer height (optional) parameter

### LINKED ISSUE #2735: "Infill combination" feature ignores layer height limitations from "printer settings" section
### OrcaSlicer Version

1.8.0-rc2

### OS version

Windows 10, Windows 11

### Additional system information

_No response_

### Printer

Kingroon KP3S (custom preset)

### How to reproduce

1. Go to "Extruder" tab in 'Printer settings"
2. Set max layer height limit less than [layer_height * n]. For example, I tried it with 0.4 max layer height limit and 0.125 layer height
3. Enable "Infill combination" feature (process settings - strength - advanced)
image

### Actual results

Maximum layer height becomes greater than layer height limit and proportional to current layer height

### Expected results

Maximum layer height should be less than maximum layer height limit

### Project file & Debug log uploads

[Layer Height Test.zip](https://github.com/SoftFever/OrcaSlicer/files/13348960/Layer.Height.Test.zip)

### Checklist of files to include

- [ ] Log file
- [X] Project file

### LINKED ISSUE #4066: Currently it is possible to enable infill combination but not to specify how many layers we want the infill combined by which is the case on superslicer
### OrcaSlicer Version

1.9.0

### OS version

windows 11

### Additional system information

Currently it is possible to enable infill combination but not to specify how many layers we want the infill combined by.

When printing at small layer heights like 0.1mm I have found that combining infill can drastically speed up print times. But by default the combined infill prints a thick infill only every 4 layers at 0.4mm thick. This has made for some poor top surface finish on some of my parts.

The above example sliced in superslicer it is possible to specify to print infill every 2nd layer for instance resulting in a 0.2mm thick infill or even every 3rd layer.

I feel that this would be benifical to specify printing every x number of layers. 1 being the default setting where combined infill is disabled

### Printer

ender 3

### How to reproduce

Currently it is possible to enable infill combination but not to specify how many layers we want the infill combined by.

### Actual results

When printing at small layer heights like 0.1mm I have found that combining infill can drastically speed up print times. But by default the combined infill prints a thick infill only every 4 layers at 0.4mm thick. This has made for some poor top surface finish on some of my parts.

### Expected results

Like in superslicer it is possible to specify to print infill every 2nd layer for instance resulting in a 0.2mm thick infill or even every 3rd layer.

### Project file & Debug log uploads

image

### Checklist of files to include

- [X] Log file
- [ ] Project file

### LINKED ISSUE #5421: Allow changing infill combination layer height
### Is there an existing issue for this feature request?

- [X] I have searched the existing issues

### Is your feature request related to a problem?

Currently infill combination height seems to be tied to nozzle width, e.g. always 0.4mm when using a 0.4mm nozzle. This makes some 3d infills unstable especially at low infill percentage. I want to combine infill every 2 layers when printing at 0.1mm layer height.

### Which printers will be beneficial to this feature?

All

### Describe the solution you'd like

Add a custom config option that changes how many layers got combined/combined infill layer height.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### LINKED ISSUE #6391: Solid Infill Combination
### Is there an existing issue for this feature request?

- [X] I have searched the existing issues

### Is your feature request related to a problem?

I was inspired by this [video](https://www.youtube.com/watch?v=ss610c_KxKc) about taking grams of plastic and printing time out of infill for strengthening parts (except for that which is strictly necessary to support top surfaces) and using it to thicken the shell instead.

I have developed a routine for simple parts to print thicker bottom and top layers quickly. My default print heigh is 0.15mm with infill combination enabled, however internal solid infill is also printed at 0.15 instead of 0.3. (0.14-0.16 is the optimal height for strength on a 0.4mm nozzle.)

So I use an object modifier that overlaps the bottom layers and top layers with settings set as follows:
- 1-2 top/bottom layer
- 100% infill
- monotonic infill pattern

This basically achieves internal solid infill printed at infill combination layer height.

It would be nice to have a checkmark to do this for all top/bottom surfaces, especially if we can avoid 1-2 overhanging surfaces so the combination layers are well supported.

One problem I anticipate is using this with layer heights <33% of the nozzle width (like 0.1mm with a 0.4mm nozzle). This will result in the infamous problem of greater than 80% nozzle width layer heights. Imagine: a layer height of 0.4mm on a 0.4mm nozzle for internal solid layers. The whole point of printing more internal solid infill for strength, and this would ruin the layer adhesion.

### Which printers will be beneficial to this feature?

All

### Describe the solution you'd like

A checkbox to enable this feature and (optionally) an integer for how many top/bottom layers should be combined away, or else how many should be retained for overhang/bridging and top surface handling.

Also, for anyone who prints with layer heights below 1/3 of the nozzle width it would be necessary to provide some sort of cap on the number of layer combinations.

### Describe alternatives you've considered

The current alternative available is to manually drop lots of modifier objects to target all the top/bottom layers you want to convert to this style of combined internal solid infill.

### Additional context

Also, #5421 here is the latest enhancement issue asking for a way to control the number/multiplier on combination layers.

### PR DESCRIPTION
**Pull request by igiannakas**
# Description

Introduced option that is only visible when combine infill is selected. It allows to set the maximum layer height over which the combination happens. Defaults to 0, which uses the nozzle diameter as it is currently done.

This setting is useful when using combine infill with layer heights that when combined result in the full diameter of the nozzle being used, resulting in weaker parts. 

For example a 0.1 layer height print using combined infill for a 0.4 nozzle would result in a 0.4 layer height infill which is fast but weak. For better layer adhesion, 80% of the nozzle width should be used (0.32) which would mean 3 layers combined instead of 4. 

Leaving this option to 100% (default) uses the nozzle diameter as the maximum, replicating the current behaviour. For user failsafe, 0 is also set to use the nozzle diameter.

**EDIT: Updated PR to also allow a % of the nozzle size for the region.**

**0.1 LH combined over 4 layers for a 0.4 nozzle**
image

**0.1 LH combined over 3 layers for a 0.4 nozzle when 0.32 infill combination - max layer height is set**
image

**Available as an option on a per object**
image

**EDIT**
Now also enabled as a % over the nozzle size used to print the region
image

# Screenshots/Recordings/Graphs

## Tests

Tested with setting applied to global settings
Tested with setting applied on a per object
Tested with 0.4 nozzle, 0.1 LH and value of 0.32 -> combined every 3 layers.
Tested with 0.4 nozzle, 0.16 LH and value of 0.32 -> combined every 2 layers.
Tested with 0.4 nozzle, 0.1LH and value of 0 -> combined every 4 layers. (ignoring this parameter and using the nozzle size as the limit - default behaviour)

Fixes #2735
Fixes #5421
Fixes #4066

## Credits:
To Noisyfox for the original implementation that pointed me in the right direction where the change needed to be made
https://github.com/SoftFever/OrcaSlicer/pull/2757

### CODE REVIEW

**Comment by igiannakas** [comment]
I'll check and see what I can do...

**Comment by igiannakas** [comment]
> Could this also be also a percentage instead of mm?

Done.

**Comment by vovodroid** [comment]
May be it's good enough to hardcode max layer height to 65-80% of infill width (not nozzle diameter)?
I use following code in my fork:

```
                float w = region.config().sparse_infill_line_width.get_abs_value(nozzle_diameter);
                if (w == 0)
                    w = this->config().line_width.get_abs_value(nozzle_diameter);

                if (current_height + layer->height >= w / 1.5 + EPSILON) {
```

**Comment by igiannakas** [comment]
But isn’t layer adhesion mostly linked to layer height? So even say with a 0.8 line width on a 0.4 nozzle a 0.4 sparse infill layer height would still be brittle, no? Haven’t tried it but that’s my instinct. 

https://help.prusa3d.com/article/layers-and-perimeters_1748#

Logic stems from here

**Comment by vovodroid** [comment]
> But isn’t layer adhesion mostly linked to layer height?

Sure. But sparse infill actually should be for "roof" support, not bending strength. So I guess some reasonable value could be hardcoded.

**Comment by CristianGTR97** [comment]
Great idea

File: src/libslic3r/PrintConfig.cpp [hunk 1]
```diff
@@ -2802,7 +2802,19 @@ void PrintConfigDef::init_fff_params()
...
     def->set_default_value(new ConfigOptionBool(false));
-
+    
+    // Orca: max layer height for combined infill
+    def = this->add("infill_combination_max_layer_height", coFloatOrPercent);
+    def->label = L("Infill combination - Max layer height");
+    def->category = L("Strength");
+    def->tooltip = L("Maximum layer height for the combined sparse infill. \n\nSet it to 0 or 100% to use the nozzle diameter (for maximum reduction in print time) or a value of ~80% to maximize sparse infill strength.\n\n"
+                     "The number of layers over which infill is combined is derived by dividing this value with the layer height and rounded down to the nearest decimal.\n\n"
+                     "Use either absolute mm values (eg. 0.32mm for a 0.4mm nozzle) or % values (eg 80%).");
```
Comment (SoftFever) [inline src/libslic3r/PrintConfig.cpp:2812]: Would it be better to mention it can't be larger than the nozzle diameter?

Comment (igiannakas) [reply src/libslic3r/PrintConfig.cpp]: Done :)

**Review by SoftFever — APPROVED** [review]
Looks good
Thank you!

**Pull request merged**